packages/flutter/lib/src/foundation/binding.dart

```diff
@@ -231,6 +236,33 @@ abstract class BindingBase {
     assert(!locked);
   }
 
+  FrameCallback _debugLastOnBeginFrame;
+  VoidCallback _debugLastOnDrawFrame;
+
+  /// Cause the entire application to stop drawing, e.g. in preparation for hot
+  /// reload.
+  ///
+  /// This is used by development tools when the application code is being
+  /// changed, to ensure that animations will not continue to allow builds to
+  /// fire before `ext.flutter.reassemble` is sent. It can be triggered
+  /// manually by sending the `ext.flutter.waitForReassemble` service extension
+  /// signal.
+  ///
+  /// This method must never be used in production code, as it will prevent all
+  /// drawing until [reassembleApplication] is called, which is also not meant
+  /// for production code. It is only meant for use in hot reload.
+  @mustCallSuper
+  @visibleForTesting
+  Future<void> waitForReassemble() async {
+    assert(() {
+      _debugLastOnBeginFrame ??= window.onBeginFrame;
+      _debugLastOnDrawFrame ??= window.onDrawFrame;
+      window.onDrawFrame = null;
+      window.onBeginFrame = null;
+      return true;
+    }());
```

Comment:
this is at the wrong layer. The foundation binding doesn't know about these IIRC.

Reply:
Window was introduced into foundation a while back now.  Are you suggesting that needs to be backed out?  I don't disagree, but it's been in there for a while.